web/packages/teleterm/src/services/appUpdater/autoUpdatesStatus.ts

```diff
@@ -0,0 +1,293 @@
+/**
+ * Teleport
+ * Copyright (C) 2025 Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import type {
+  ClusterVersionInfo,
+  GetClusterVersionsResponse,
+  UnreachableCluster,
+} from 'gen-proto-ts/teleport/lib/teleterm/auto_update/v1/auto_update_service_pb';
+import { compare, major } from 'shared/utils/semVer';
+
+import Logger from 'teleterm/logger';
+import { RootClusterUri } from 'teleterm/ui/uri';
+
+const logger = new Logger('resolveAutoUpdatesStatus');
+
+/**
+ * Determines the auto-update status based on all connected clusters.
+ * Updates can be either enabled or disabled, depending on the passed sources.
+ * The version is selected using the following precedence:
+ * 1. `TELEPORT_TOOLS_VERSION` env var, if defined.
+ * 2. `tools_version` from the cluster manually selected to manage updates, if selected.
+ * 3. Most compatible version, if found.
+ * 4. If there's no version at this point, stop auto updates.
+ */
+export async function resolveAutoUpdatesStatus(sources: {
```

Comment:
I assume we'll be able to test it once the UI for this is ready, right?

There might be some other cases we haven't thought of, but I think it can become more apparent once we're able to actually test this in a real app.

Reply:
>I assume we'll be able to test it once the UI for this is ready, right?

Yes, in the upcoming PR I'll build the UI based on the `AppUpdateEvent`. You'll be able to play with it in the Storybook. Then, in the next PR, I'll connect the UI with the actual updater.